### ISSUE TITLE
Allow backup of NVROM files for which their are data offsetds for games

### ISSUE THREAD

**Issue by mmattner** [mmattner/vPinBackupManager#10]
A request from Syd to support integration with VPin Studio is the abiltiy to backup data from NVRAM files with offsets - see below:

--------------------------------------------
NVRam

Scripts can have an nvOffset to support different highscore files for the same table. This is used when tables have been modded, but you want to differ their highscores. As a result, you will have different nvram that do not necessarily match with the ROM name / gamecode. I'm just curious if you support that.
---------------------------------------------

Ive asked Syd for an example of games that use this.
Once done the idea will be to determine how the offsrt is used, then at startup "extract" the appropriate set of data at the correct offset and store in the temp locations ... I do this for VPReg.stg content so can advise. Once this is done its easier to do diffs of it. We then need code to be able to re-insert it .. .the challenge here will be if multiple games use the same offset.

**Comment by linojon** [comment]
ok. Verified. heres how it works:
VPX lets you specify a "version" or "offset" NVRAM file in the table script by adding the line `NVOffset n` where n is a number between 1 and 32. At runtime it creates an empty txt file as a "bookmark" indicating which version the current [GAMECODE].nv file is. When you run a game, if the bookmark is the same as the NVOffset in the script nothing happens. If its different, it renames the bookmark for the new version, and swaps the .nv files. For example:

- I added "NVOffset 2" to a game (must be after the "GameName = " line), run table, it creates an empty .txt file as a bookmark, e.g. acd_170h v2.txt. Theres still one .nv file for this rom
![image](https://user-images.githubusercontent.com/68624/234736299-7218ab8c-1185-477a-8445-26695c10a3eb.png)

- quit game, change it to `NVOffset 3` , run table. It renames the bookmark to  "acd_170h v3.txt" backs up the previous .nv file creating a file "acd_170h v2.nv". The regular .nv file is now considered to be the v3 one
![image](https://user-images.githubusercontent.com/68624/234736335-6561eef9-61d0-4a08-935a-a323713be0ef.png)

- quit game, change it back to `NVOffset 2`, run table. Renames the bookmark to "acd_170h v2.txt", backs up the acd_170h.nv file to "acd_170h v3.nv", and copies “acd_170h.v2.nv” to “acd_170h.nv” to run the game with that version.
![image](https://user-images.githubusercontent.com/68624/234736357-d62aadfa-3c77-4e9f-bfc8-5e8c38e837e1.png)

**Comment by mmattner** [comment]
Great analysis.  So it looks like we won't need to backup the bookmark file at all as it's only an indication of which version/offset the current NVRAM file is.

So logic would be something like:
- check if there is an NVRAM offset, if so store it. 
- check NVRAM folder and see if bookmark matches this games offset, if so copy the default NVRAM file to backup location, if not copy appropriate backup version. 

I think it's best we do this up front so we can just grab the md5 of the copied file. 

NVRAM filters become a special case - so remove existing filter and add new one with a new type NVRAM

The major operations then are:

Backup
- check if a temp NVRAM was copied to temp spot at startup if so copy that to backup folder and proceed as normal

Delete 
- delete backups and txt files plus the NVRAM file itself unless we detect another game using same ROM without an NVRAM offset

Restore
- copy NVRAM into correct folder but store as the backup version only. System will grab it when needed. 

The processing in the three processes above does need to be in its own if block like the VPReg stuff for now - but we will find a way to tidy this up over time if needed

**Comment by mmattner** [comment]
In terms of a filter, the folliwing changes are required:
- Modify line 21 of Constants.cs (and associated comment) to be:  public enum FILTER_PATH_TYPE { FILE, NVRAM,  VPREG };

- In Managers/SettingsManager.cs at around line 289, insert: 
new PathFilter(@"VPinMAME\nvram", new List<string> { Constants.TOKEN_GAMECODE + " v[NVRAMOFFSET].nvram", Constants.TOKEN_GAMECODE + " v[NVRAMOFFSET].nv" }, "NVRAM (non-volatile RAM) file", false, FILTER_PATH_TYPE.NVRAM),

At this point in time the filter will find both the default NV files PLUS any offset backups. At this point we need to add special processing into the backup/delete/restore code - all should sit inside the block iterating through filters (next to where the checks for VPREG are) Like below:
![image](https://user-images.githubusercontent.com/483200/234743599-dd19dcde-80fe-4e84-b07f-a66bf87b9570.png)

In this block add an "if (pathFilters.Type == FILTER_PATH_TYPE.NVRAM)" - inside this block will be the special processing.
The processing for a backup should be:
- if a file exists (ie a backed up NVRAM) then use that in the backup
- if no file exists (ie no backed up NVRAM) then check if this is the only game using the ROM, if it is, use the default MVRAM file (if it exists) in backup, otherwise use nothing
Store the file in the backup as the default NVRAM name, we will adjust on restore, but that way it will always match on MD5s regardless of offset etc

The processing for delete should be:
i delete any text file and backup file that exists, if no other games use the ROM, also delete the main NVRAM file

The processing for restore should be:
- check if the table file is using an offset, if it is, then check if that ofset is being used by ANOTHER table (ie the case where a game had an offset, was deleted, then replaced by another table with the same offset). - if there is a conflict  WARN the user i f not just copy the file to the correct backup name


## Let me know if you think Ive missed anything ?

**Comment by linojon** [comment]
Here's some additional edge cases. 

* If the table uses NVOffset and is backed up, and the user plays a game with a different version so the live rom is a different version. Then you restore this one, you want the rom to go into the versioned nv file, not the live one..
* If you delete a table with NVOffset, should only delete its versioned nv file. May be best to leave the live .nv file alone in case a game uses it but doesnt 
* Not everyone knows or cares about NVOffset so they might have multiple tables (eg versions or mods) that share the rom. You dont want to accidentally delete or even restore the nv file, and the rom itself (and any other files based on the rom name for that matter) that is shared
* its also possible for two games to specify the same NVOffset version number. So then delete and restore will mess that up too.
 
 I suggest:
 
**Backup**
* if the table script contains NVOffset, get the version number
* else back up the live .nv file itself and done.
* if the current bookmark file is for this version, backup the live .nv file but name it with the versioned name in our backup ([GAMECODE].vN.nv)
* else back up the version named file.
* Thus the backup will always contain a file with the versioned name if it uses NVOffset

**Delete**
* if another game uses the same .nv file in backup (versioned or live), dont delete the nv file 

Restore
* restore the live or versioned nv file, whichever is in the backup
* [if another table uses this rom, and this is not versioned, ask the user and/or suggest they add NVOffset to their tables]

**Comment by linojon** [comment]
Pushed changes that i think now implements this correctly in BackupManager.

The backup algorithm is as follows:
Given:
 activeFilename is the actual live nvram file, eg "<romname>.nv"
versionFilename iswhat this filter has identified eg "<romname> v2.nv", may or may not exist
bookmarkFilename is based on filter eg "<romname> v2.txt", may or may not exist

Algorithm:
if the activeFilename does not exist, ignore this filter
else if theres a bookmark, backup the active one (to versioned name)
else if theres a versioned one, back that up
else nothing to do (active one has its own filter already)

todo: 

- need to extract the table's NVOffset value from the table script. For development uncomment line 1108 (+/-) and give it the version number (as a string).
- delete: dont remove the nv/nvram file if used by another table
- (restore should be ok without changes)
- 
also, now when a backup is completed, the UI shows it in red. Havent investigated this.

![image](https://user-images.githubusercontent.com/68624/235281130-6a378bd3-889d-4498-b7ac-6f6a6c2cb030.png)

**Comment by linojon** [comment]
Discovered another issue. The manifest should contain only one .nv (or .nvram) file. So the regular .nv filter will also be tagged at FILTER_PATH_TYPE.NVRAM and the special cases handled in the code.

**Comment by mmattner** [comment]
> Discovered another issue. The manifest should contain only one .nv (or .nvram) file. So the regular .nv filter will also be tagged at FILTER_PATH_TYPE.NVRAM and the special cases handled in the code.

Yes - this is what I alluded to with the need to consider deleting the regular filter match IF an offset filter hit. The exact implementation I cant give as there multiuple ways of doing this, specifying them both as NVRAM filters is probably one way.


You may have done it already, but we also need to make sure it works in the following case you listed above:
![image](https://user-images.githubusercontent.com/483200/235408031-654c2afd-4d4a-4699-bd4f-9e99c15b65c4.png)

Ie - the filter wont hit on the rom.v2.nv file as it doesnt exist yet. But in this case we need to grab the default rom.

**Comment by linojon** [comment]
Everything seems to be working but requires Safe mode to be enabled for shared rom files (out of scope of this issue).
I suggest leaving this open until that is resolved.
